### PR TITLE
Allow to pass explicit `CostTable` for the custom native functions

### PR DESCRIPTION
**Pull request by mkurnikov**
## Motivation

Allows external users of Move language to extend the set of native functions without forking the repo. 

This implementation is not ideal as one would have to copy `NativeCostIndex` and remember to start with `18` in the enum values. It would be better to make an extensible `NativeCostIndex` enum somehow or replace it, but I've decided it's too big of a change for a first contribution. I can follow up with that later. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

File: language/move-vm/types/src/natives/function.rs [hunk 1]
```diff
@@ -106,12 +106,12 @@ impl NativeResult {
 
 /// Return the native gas entry in `CostTable` for the given key.
 /// The key is the specific native function index known to `CostTable`.
-pub fn native_gas(
+pub fn native_gas<NativeIndex: Into<u8>>(
     table: &CostTable,
-    key: NativeCostIndex,
+    native_table_idx: NativeIndex,
```
Comment (vgao1996) [inline language/move-vm/types/src/natives/function.rs:111]: Change this to `native_table_idx: impl Into<u8>` to make the function signature a bit cleaner?

**Review by vgao1996 — COMMENTED** [review]
Seems reasonable to me. A few nits:

File: language/move-vm/types/src/gas_schedule.rs [hunk 2]
```diff
@@ -452,11 +457,11 @@ pub fn calculate_intrinsic_gas(
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 #[repr(u8)]
 pub enum NativeCostIndex {
-    SHA2_256 = 0,
-    SHA3_256 = 1,
-    ED25519_VERIFY = 2,
-    ED25519_THRESHOLD_VERIFY = 3,
-    BCS_TO_BYTES = 4,
+    BCS_TO_BYTES = 0,
```
[the remark this answers is not in the document]
Comment (mkurnikov) [reply language/move-vm/types/src/gas_schedule.rs:460]: Shuffling the indices to support a case where this `unwrap_or_else` invocation happens https://github.com/diem/move/blob/main/language/tools/move-unit-test/src/test_runner.rs#L132-L134
and only move natives are loaded. There are only 16 native functions then, and when `EMIT_EVENT = 17` is invoked test crashes. 
I haven't found any requirements for those indices across the codebase aside from being consecutive and starting from zero, so I reshuffled Diem-only native indices to the end.

[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/types/src/gas_schedule.rs:460]: @tzakian will need to comment, but I don't think these can be shuffled at all? Don't they end up in cost tables that might be serialized in storage somewhere?

[the remark this answers is not in the document]
Comment (tzakian) [reply language/move-vm/types/src/gas_schedule.rs:460]: Todd's correct. The cost table for natives is stored as `Vec<GasCost>` that is indexed by this `NativeCostIndex`. So any changes in the value assigned here would need to be in-synch with changes to the published native gas schedules that are stored on chain.

The only solution that comes to mind right now is that the gas schedule will probably need to be padded for the Diem-specific segment (so I believe the num_of_native_funcs value will need to be a minimum of 17 always)

[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-vm/types/src/gas_schedule.rs:460]: Maybe if/when we have maps we could switch this table to a map?

I guess regardless I'm not really sure why you need to fork the repo in the first place?

[the remark this answers is not in the document]
Comment (mkurnikov) [reply language/move-vm/types/src/gas_schedule.rs]: Thank you all for the review, I've removed the indices shuffling and instead added vector length padding to `zero_cost_schedule()` as @tzakian suggested. 

> I guess regardless I'm not really sure why you need to fork the repo in the first place?

I haven't found any other way to provide custom `CostTable` for our own native functions in the `run_cli` call. There's also no way to provide a custom boogie prelude for those native functions, but it's a feature for the next time.

**Review by vgao1996 — APPROVED** [review]
LGTM

**Pull request closed without merging**